plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -188,9 +208,27 @@ public void put(final MavenPath path, final Payload payload)
     try (StorageTx tx = getStorage().openTx()) {
       if (path.getCoordinates() != null) {
         putArtifact(path, payload, tx);
+        if (payload instanceof MavenPayload) {
```

Comment:
Couple of places we cast but do nothing if we are passed the wrong type, and I'm thinking that if we get here and have something other than a MavenPayload, it should be an error condition.